app/instrumentation-tests/src/org/commcare/androidTests/LanguagesTest.kt

```diff
@@ -0,0 +1,135 @@
+    }
+
+    @Before
+    fun setup() {
+        if (CommCareApplication.instance().currentApp == null) {
+            InstrumentationUtility.installApp(FormEntryTest.CCZ_NAME)
+        } else {
+            InstrumentationUtility.uninstallCurrentApp()
+            InstrumentationUtility.installApp(FormEntryTest.CCZ_NAME)
+            Espresso.pressBack()
+        }
+        InstrumentationUtility.login("user_with_no_data", "123")
+    }
+
+    @Test
+    fun testLanguageChanges() {
+        InstrumentationUtility.openForm(0, 0)
+        onView(withText("Enter a name:"))
+                .check(matches(isDisplayed()))
+        exitForm()
+        InstrumentationUtility.gotoHome()
+        selectLanguageChangeOption()
+        checkLanguageOptions()
+
+        // Confirm the options persist on rotation
+        InstrumentationUtility.rotateLeft()
+        checkLanguageOptions()
+        InstrumentationUtility.rotatePortrait()
+        checkLanguageOptions()
+
+        // Clicking just after rotating doesn't work for some reason, so waiting here for a moment.
+        InstrumentationUtility.sleep(2)
+        onView(withText("Hindi"))
+                .perform(click())
+        onView(withText("Start"))
+                .perform(click())
+        onView(withText("Basic Form Tests"))
+                .perform(click())
+        onView(withText("HIN: Languages"))
+                .perform(click())
+
+        onView(withText("HIN: Enter a name:"))
+                .check(matches(isDisplayed()))
+
+        // Check form language changes
+        selectLanguageChangeOption()
+        checkLanguageOptions()
+        onView(withText("English"))
+                .perform(click())
+
+        onView(withText("HIN: Enter a name:"))
+                .check(doesNotExist())
+        onView(withText("Enter a name:"))
+                .check(matches(isDisplayed()))
+
+        // Confirm app language remains the same.
+        exitForm()
+        onView(withText("HIN: Languages"))
+                .check(matches(isDisplayed()))
+    }
+
+    @Test
+    fun testAppUpdate_dontInterfereLanguageChanges() {
+        selectLanguageChangeOption()
+        checkLanguageOptions()
+        // Change language to hindi
+        onView(withText("Hindi"))
+                .perform(click())
+
+        // Update app
+        InstrumentationUtility.openOptionsMenu()
+        onView(withText("Update App"))
+                .perform(click())
+        onView(withText("Update to version 103 & log out"))
+                .perform(click())
+        InstrumentationUtility.login("user_with_no_data", "123")
+
+        // Confirm app language remains hindi
+        onView(withText("Start"))
+                .perform(click())
+        onView(withText("Basic Form Tests"))
+                .perform(click())
+        onView(withText("HIN: Languages"))
+                .check(matches(isDisplayed()))
+    }
+
+    private fun selectLanguageChangeOption() {
```

Comment:
This can be generalized and move to `InstrumentationUtility` as `selectMenu(int textId)`,